src/Polysemy/Internal/Union.hs

```diff
@@ -51,10 +58,10 @@ import Polysemy.Internal.CustomErrors
 data Union (r :: EffectRow) (m :: Type -> Type) a where
   Union
       :: -- A proof that the effect is actually in @r@.
-         SNat n
+         ElemOf r e
          -- The effect to wrap. The functions 'prj' and 'decomp' can help
          -- retrieve this value later.
-      -> Weaving (IndexOf r n) m a
+      -> Weaving e m a
```

Comment:
It's not immediately obvious to me how this works without the `IndexOf`. IIUC `ElemOf` is a proof that `e` is in `r`, but nothing constrains the `e` in `Weaving` to be where we think it is?

Reply:
Nevermind, I get it.